File: 0092.reverse_linked_list_II/solution.py
from utils import ListNode


class Solution:
    def reverse_between(self, head: ListNode, m: int, n: int) -> ListNode:
        if not head:
            return head
        ans = ListNode(None)
        ans.next = head
        p = ans
        for _ in range(m - 1):
            p = p.next

        head = p.next
        for _ in range(m, n):
            temp = head.next
            head.next = temp.next
            temp.next = p.next
            p.next = temp
        return ans.next

File: 0092.reverse_linked_list_II/solution.go
package leetcode0092

import "github.com/Ackerr/Algorithms/utils"

func reverseBetween(head *utils.ListNode, m int, n int) *utils.ListNode {
	if head == nil || m > n {
		return head
	}
	ans := &utils.ListNode{}
	ans.Next = head
	p := ans
	for i := 1; i < m; i++ {
		p = p.Next
	}
	head = p.Next
	for i := m; i < n; i++ {
		temp := head.Next
		head.Next = temp.Next
		temp.Next = p.Next
		p.Next = temp
	}
	return ans.Next
}
